core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,208 @@
+package ca.jrvs.apps.grep;
+
+import java.io.*;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Collectors;
+
+
+public class JavaGrepImp implements JavaGrep{
+
+    private String regex;
+    private String rootPath;
+    private String outFile;
+
+    public static void main(String[] args) {
+        if (args.length != 3) {
+            throw new IllegalArgumentException("USAGE: regex rootPath outFile");
+        }
+
+        JavaGrepImp javaGrepImp = new JavaGrepImp();
+        javaGrepImp.setRegex(args[0]);
+        javaGrepImp.setRootPath(args[1]);
+        javaGrepImp.setOutFile(args[2]);
+
+        try {
+            javaGrepImp.process();
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
+
+    }
+
+    @Override
+    public void process() throws IOException {
+
+        List<File> files = this.listFiles(rootPath);
+        List<String> matchedLines =  new ArrayList<>();;
+
+        for(File file : files){
+            List<String> lines = this.readLines(file);
+            boolean matchFound = false;
+            for(String line: lines){
+                matchFound = this.containsPattern(line);
+
+                //check if match is found
+                if(matchFound){
+                    matchedLines.add(line);
+                }
+            }
+        }
+
+        //check if matched Lines found and write to file
+        if(matchedLines!=null){
+            this.writeToFile(matchedLines);
+            System.out.println("Patterns has been written to File");
+        }else{
+            System.out.println("Pattern Not found");
+        }
+
+    }
+
+    //traverse a directory and return all the files
+    @Override
+    public List<File> listFiles(String rootDir)
+    {
+        File directory = new File(rootDir);
+
+        // Check if rootDir is Directory
+        if(!directory.exists() || !directory.isDirectory()){
+            throw new IllegalArgumentException(String.format("%s is not a directory", rootDir));
+        }
+
+        List<File> fileList = new ArrayList<>();
+        File rootdir = new File(rootDir);
+
+        //Get the name of directories contained in the dir
+        if (rootdir.isDirectory()) {
+            File[] files = rootdir.listFiles();
+
+
+        if (files != null) {
```

Comment:
This line is unnecessary.  When running a for each loop if the collection is empty it will not run.